mountpoint-s3/src/data_cache/express_data_cache.rs

```diff
@@ -94,17 +119,36 @@ where
             return Err(DataCacheError::InvalidBlockOffset);
         }
 
-        let object_key = block_key(&self.prefix, cache_key, block_idx);
+        let object_key = get_s3_key(&self.prefix, cache_key, block_idx);
         let result = match self
             .client
-            .get_object(&self.bucket_name, &object_key, &GetObjectParams::new())
+            .get_object(
+                &self.bucket_name,
+                &object_key,
+                &GetObjectParams::new().checksum_mode(Some(ChecksumMode::Enabled)),
+            )
             .await
         {
             Ok(result) => result,
             Err(ObjectClientError::ServiceError(GetObjectError::NoSuchKey)) => return Ok(None),
             Err(e) => return Err(e.into()),
         };
 
+        let object_metadata = result
+            .get_object_metadata()
+            .await
+            .map_err(|err| DataCacheError::IoFailure(err.into()))?;
+
+        let checksum = result
+            .get_object_checksum()
+            .await
+            .map_err(|err| DataCacheError::IoFailure(err.into()))?;
+        let crc32c = crc32c_from_base64(&checksum.checksum_crc32c.ok_or(DataCacheError::BlockChecksumMissing)?)
+            .map_err(|err| DataCacheError::IoFailure(err.into()))?;
+
+        let block_metadata = BlockMetadata::new(block_idx, block_offset, cache_key, &self.bucket_name, crc32c);
```

Comment:
This is a bit non-symmetric, that all of the arguments to `BlockMetadata::new` are from the "expected" parameters, passed to `get_block`, with the exception of  `crc32c` which is read from the actual block. But I think this is ok, since having block's crc32 in metadata's checksum mitigates another risk.

Reply:
I could migrate the `crc32c` field to be arguments to both `to_put_object_params` and `validate_object_metadata`, but I feel that would probably be worse

Reply:
I think it's ok, the main part is that we actually compare other params to the expected values.